### ISSUE TITLE
isStage1Descriptor: false positive, if passed a function

### ISSUE THREAD

**Issue by buschtoens** [ember-decorators/ember-decorators#327]
If you pass a `function() {}` as the only argument to a decorator, [`isStage1Descriptor`](https://github.com/ember-decorators/ember-decorators/blob/f6259d7e38e25f60d249e8771f340b89e50dc42f/packages/utils/addon/decorator.js#L15-L38) returns a false positive.

```ts
// handler is a function
const myDecorator = decoratorWithRequiredParams((desc, [handler]) => desc);

// breaks
@myDecorator(function() {})
class Foo {}

// works
@myDecorator(() => {})
class Foo {}
```

https://github.com/ember-decorators/ember-decorators/blob/f6259d7e38e25f60d249e8771f340b89e50dc42f/packages/utils/addon/decorator.js#L31-L35

Do you think that there is any way to reliably tell a `function() {}` apart from a `class {}`?

Funny coincidence, that ["How Does React Tell a Class from a Function?" by Dan Abramov](https://overreacted.io/how-does-react-tell-a-class-from-a-function/) is trending on HN now. 😄

**Comment by buschtoens** [comment]
Taking the [babel class transforms](https://babeljs.io/en/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=true&code_lz=MYewdgzgLgBAYjAvDYAbAhhCMDeAoGGAWwFMoALEAEwAoBKXAXz2b1EhFRIDpUQBzGnDpA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=es2015%2Ces2016%2Ces2017%2Cstage-0%2Cstage-1%2Cstage-2%2Cstage-3%2Ces2015-loose%2Ctypescript&prettier=false&targets=&version=7.0.0-beta.50&envVersion=7.0.0-beta.51) into account, I don't think there is any way to tell the difference between a `function() {}` and a `class {}`.

We could either do it on a best effort basis, which will still fail in certain cases, or just outright "ban" users from passing a single function as a parameter.

Some ideas for heuristics:

- We can check the length of the prototype chain. `function() {}` and `class {}` have the same length, but derived classes have longer chains: One extra link per `extends`.
- We can check whether `target.prototype instanceof Ember.Object`. This would reliably detect all classes that extend Ember framework classes.

We could also ship an _extra_ babel transform, that transforms all decorator invocations with one parameter singular parameter to be "tagged".

```js
@myDecorator(singleParam)
class Foo {}
```

```js
import { tag } from '@ember-decorators/utils/tag';

@myDecorator(tag(singleParam))
class Foo {}
```

```js
const tagged = new WeakSet();

export function tag(param) {
  if (typeof param === 'function') {
    tagged.add(param);
  }
  return param;
}

export function isTagged(param) {
  return typeof param === 'function' || tagged.has(param);
}
```

This would be 100 % safe. Is it overkill?

Do we only need this transform for Babel 6 or also for Babel 7?

**Comment by buschtoens** [comment]
If anyone also runs into this, these are the ways out:

- Use an arrow function instead.
- Pass a second, to be ignored argument, like `undefined`, to defuse the `arguments.length === 1` check.
- Strip off the `prototype` from the function before you pass it.

**Comment by pzuraq** [comment]
Yeah, not sure if it's really worth the effort given stage 1 decorators are on the way out. I think passing an arrow function is the best idea, though that could be tricky if the function needs to have `this` rebound.
